rct229/rulesets/ashrae9012019/ruleset_functions/compare_swh_dist_systems_and_components.py

```diff
@@ -0,0 +1,133 @@
+from rct229.rulesets.ashrae9012019.data_fns.extra_schema_fns import (
+    EXTRA_SCHEMA,
+    compare_context_pair,
+)
+from rct229.rulesets.ashrae9012019.ruleset_functions.get_swh_equipment_associated_with_each_swh_distriubtion_system import (
+    get_swh_equipment_associated_with_each_swh_distribution_system,
+)
+from rct229.utils.assertions import assert_
+from rct229.utils.utility_functions import (
+    find_exactly_one_pump,
+    find_exactly_one_service_water_heating_distribution_system,
+    find_exactly_one_service_water_heating_equipment,
+)
+
+
+def compare_swh_dist_systems_and_components(
+    rmd1: dict, rmd2: dict, compare_context_str: str, swh_distribution_id: str
+):
+    """This function compares all sub-components of a given SWH Distribution System ID between two models.
+    Parameters
+    ----------
+    rmd1: dict
+        RMD at RuleSetModelDescription level
+    rmd2: dict
+        RMD at RuleSetModelDescription level
+    compare_context_str: str
+        compare context value from the extra schema (e.g., RMD Test, AppG Used By TCDs, AppG P_RMD Equals U_RMD, AppG B_RMD Equals P_RMD, AppG B_RMDs Same)
+    swh_distribution_id: str
+        id of the `service_water_heating_distribution_systems` key to compare
+
+    Returns
+    -------
+    errors: list
+        An array of errors encountered in this function and in compare_context_pair. If errors is an empty array, then all elements that were expected to match do match
+
+    """
+
+    errors_list = []
+    rmd1_swh_system_and_equip_dict = (
+        get_swh_equipment_associated_with_each_swh_distribution_system(rmd1)
+    )
+    rmd2_swh_system_and_equip_dict = (
+        get_swh_equipment_associated_with_each_swh_distribution_system(rmd2)
+    )
+
+    # service_water_heating_distribution_systems
+    assert_(
+        rmd1_swh_system_and_equip_dict[swh_distribution_id],
+        f"SWH Distribution System {swh_distribution_id} not found in the `rmd1`.",
+    )
+    assert_(
+        rmd2_swh_system_and_equip_dict[swh_distribution_id],
+        f"SWH Distribution System {swh_distribution_id} not found in the `rmd2`.",
+    )
+    rmd1_swh_equipment_dict = rmd1_swh_system_and_equip_dict[swh_distribution_id]
+    rmd1_swh_distribution = find_exactly_one_service_water_heating_distribution_system(
+        rmd1, swh_distribution_id
+    )
+    rmd2_swh_equipment_dict = rmd2_swh_system_and_equip_dict[swh_distribution_id]
+    rmd2_swh_distribution = find_exactly_one_service_water_heating_distribution_system(
+        rmd2, swh_distribution_id
+    )
+
+    index_content_dict, compare_context_dict = (
+        (rmd1_swh_distribution, rmd2_swh_distribution)
+        if len(rmd1_swh_distribution) > len(rmd2_swh_distribution)
+        else (rmd2_swh_distribution, rmd1_swh_distribution)
+    )
```

Comment:
Checking the logic for getting the `rmd1_swh_distribution` and `rmd2_swh_distribution`, it does not make sense to check the length of the these two objects.
I think this logic can be skipped - index -> rmd1, compare -> rmd2

Reply:
Good point. Addressed.